src/datanode/src/sql/delete.rs

```diff
@@ -0,0 +1,139 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use common_query::Output;
+use common_telemetry::warn;
+use datatypes::prelude::VectorRef;
+use datatypes::vectors::{BooleanVector, Float64Vector, StringVector, TimestampMillisecondVector};
+use session::context::QueryContextRef;
+use snafu::ResultExt;
+use sql::ast::{BinaryOperator, Expr, Value};
+use sql::statements::delete::Delete;
+use table::engine::TableReference;
+use table::requests::DeleteRequest;
+
+use crate::error::{DeleteSnafu, InvalidSqlSnafu, Result};
+use crate::instance::sql::table_idents_to_full_name;
+use crate::sql::{SqlHandler, SqlRequest};
+
+impl SqlHandler {
+    pub(crate) async fn delete(&self, req: DeleteRequest) -> Result<Output> {
+        let table_ref = TableReference {
+            catalog: &req.catalog_name.to_string(),
+            schema: &req.schema_name.to_string(),
+            table: &req.table_name.to_string(),
+        };
+
+        let table = self.get_table(&table_ref)?;
+
+        let affected_rows = table.delete(req).await.with_context(|_| DeleteSnafu {
+            table_name: table_ref.to_string(),
+        })?;
+
+        Ok(Output::AffectedRows(affected_rows))
+    }
+
+    pub(crate) fn delete_to_request(
+        &self,
+        stmt: Delete,
+        query_ctx: QueryContextRef,
+    ) -> Result<SqlRequest> {
+        let (catalog_name, schema_name, table_name) =
+            table_idents_to_full_name(stmt.table_name(), query_ctx)?;
+        let key_column_values = parser_selection(stmt.selection())?;
+        Ok(SqlRequest::Delete(DeleteRequest {
+            key_column_values,
+            catalog_name,
+            schema_name,
+            table_name,
+        }))
+    }
+}
+
+/// parse selection, currently supported format is `tagkey1 = 'tagvalue1' and tagkey2 = 'tagvalue2'`.
+/// (only uses =, and in the where clause and provides all columns needed by the key.)
+fn parser_selection(selection: &Option<Expr>) -> Result<HashMap<String, VectorRef>> {
+    let mut key_column_values = HashMap::new();
+    if let Some(expr) = selection {
+        parser_expr(expr, &mut key_column_values)?;
+    }
+    Ok(key_column_values)
+}
+
+fn parser_expr(expr: &Expr, key_column_values: &mut HashMap<String, VectorRef>) -> Result<()> {
+    // match BinaryOp
+    if let Expr::BinaryOp { left, op, right } = expr {
+        match (left.deref(), op, right.deref()) {
+            // match And operator
+            (Expr::BinaryOp { .. }, BinaryOperator::And, Expr::BinaryOp { .. }) => {
+                parser_expr(left.deref(), key_column_values)?;
```

Comment:
Does the compiler support auto-deref this?
```suggestion
                parser_expr(&left, key_column_values)?;
```

Reply:
<img width="1505" alt="image" src="https://user-images.githubusercontent.com/44310040/217988733-3bde98c6-8971-46bb-bf73-23a1470a066a.png">

It seems that the compiler will report an error.